File: design-system/components/buttons/Button.js
/**
 * Button Component
 *
 * A reusable button component with multiple variants and sizes.
 *
 * Usage:
 * ```
 * <Button variant="primary" size="md" onClick={handleClick}>Click Me</Button>
 * <Button variant="secondary" size="sm">Secondary Button</Button>
 * <Button variant="outline" size="lg" disabled>Disabled Button</Button>
 * <Button variant="link">Link Button</Button>
 * ```
 */

class Button {
  constructor() {
    this.baseClasses =
      "flex items-center justify-center font-medium transition-colors duration-150 focus:outline-none focus:ring-2 focus:ring-offset-2";
  }

  /**
   * Renders a button with the specified variants and props
   * @param {Object} options - Button options
   * @param {string} options.variant - Button variant (primary, secondary, outline, link)
   * @param {string} options.size - Button size (sm, md, lg)
   * @param {boolean} options.disabled - Whether the button is disabled
   * @param {string} options.className - Additional classes to apply
   * @param {Function} options.onClick - Click handler function
   * @param {string} options.type - Button type (button, submit, reset)
   * @param {boolean} options.fullWidth - Whether the button should take up full width
   * @param {string} options.text - Button text
   * @param {string} options.iconLeft - Icon to show before text
   * @param {string} options.iconRight - Icon to show after text
   * @returns {string} HTML markup for the button
   */
  render({
    variant = "primary",
    size = "md",
    disabled = false,
    className = "",
    onClick = null,
    type = "button",
    fullWidth = false,
    text = "",
    iconLeft = "",
    iconRight = "",
  }) {
    // Determine variant-specific classes
    let variantClasses = "";
    let disabledClasses = "";

    if (disabled) {
      disabledClasses = "opacity-50 cursor-not-allowed";
    }

    switch (variant) {
      case "primary":
        variantClasses =
          "bg-cyan-600 hover:bg-cyan-700 text-white border border-transparent rounded-lg shadow-sm focus:ring-cyan-500";
        break;
      case "secondary":
        variantClasses =
          "bg-white hover:bg-gray-50 text-gray-700 border border-gray-300 rounded-lg shadow-sm focus:ring-cyan-500";
        break;
      case "outline":
        variantClasses =
          "bg-transparent hover:bg-cyan-50 text-cyan-600 border border-cyan-600 rounded-lg focus:ring-cyan-500";
        break;
      case "link":
        variantClasses =
          "bg-transparent text-cyan-600 hover:text-cyan-500 border-0 shadow-none focus:ring-transparent";
        break;
      default:
        variantClasses =
          "bg-cyan-600 hover:bg-cyan-700 text-white border border-transparent rounded-lg shadow-sm focus:ring-cyan-500";
    }

    // Determine size-specific classes
    let sizeClasses = "";
    switch (size) {
      case "sm":
        sizeClasses = "px-3 py-1.5 text-xs";
        break;
      case "md":
        sizeClasses = "px-4 py-2.5 text-sm";
        break;
      case "lg":
        sizeClasses = "px-5 py-3 text-base";
        break;
      default:
        sizeClasses = "px-4 py-2.5 text-sm";
    }

    // Handle full width
    const widthClass = fullWidth ? "w-full" : "";

    // Combine all classes
    const buttonClasses = [
      this.baseClasses,
      variantClasses,
      sizeClasses,
      widthClass,
      disabledClasses,
      className,
    ]
      .filter(Boolean)
      .join(" ");

    // Create icon elements if needed
    const leftIconHtml = iconLeft
      ? `<span class="mr-2">${iconLeft}</span>`
      : "";
    const rightIconHtml = iconRight
      ? `<span class="ml-2">${iconRight}</span>`
      : "";

    // Build the button HTML
    return `
      <button
        type="${type}"
        class="${buttonClasses}"
        ${disabled ? "disabled" : ""}
        ${onClick ? `onclick="${onClick}"` : ""}
      >
        ${leftIconHtml}
        ${text}
        ${rightIconHtml}
      </button>
    `;
  }

  /**
   * Creates a primary button
   */
  primary(options = {}) {
    return this.render({ variant: "primary", ...options });
  }

  /**
   * Creates a secondary button
   */
  secondary(options = {}) {
    return this.render({ variant: "secondary", ...options });
  }

  /**
   * Creates an outline button
   */
  outline(options = {}) {
    return this.render({ variant: "outline", ...options });
  }

  /**
   * Creates a link button
   */
  link(options = {}) {
    return this.render({ variant: "link", ...options });
  }
}

// Export the Button class
export default new Button();

File: app/components/button.py
"""
Button component that adheres to the shastho design system.
"""

def primary(text, type="button", size="md", full_width=False, icon=None, additional_classes=""):
    """
    Generate HTML for a primary button.

    Args:
        text (str): Button text
        type (str): Button type (button, submit, reset)
        size (str): Button size (sm, md, lg)
        full_width (bool): Whether the button should take full width
        icon (str): Optional HTML for an icon
        additional_classes (str): Additional CSS classes

    Returns:
        str: HTML for the button
    """
    # Base classes
    classes = "bg-primary hover:bg-primary-hover text-white rounded-lg font-medium transition-colors"

    # Size classes
    size_classes = {
        "sm": "px-3 py-1.5 text-sm",
        "md": "px-4 py-2.5",
        "lg": "px-6 py-3 text-lg"
    }

    classes += f" {size_classes.get(size, size_classes['md'])}"

    # Width class
    if full_width:
        classes += " w-full"

    # Add any additional classes
    if additional_classes:
        classes += f" {additional_classes}"

    # Add icon if provided
    icon_html = f"{icon} " if icon else ""

    return f'<button type="{type}" class="{classes}">{icon_html}{text}</button>'

def secondary(text, type="button", size="md", full_width=False, icon=None, additional_classes=""):
    """
    Generate HTML for a secondary button.

    Args:
        text (str): Button text
        type (str): Button type (button, submit, reset)
        size (str): Button size (sm, md, lg)
        full_width (bool): Whether the button should take full width
        icon (str): Optional HTML for an icon
        additional_classes (str): Additional CSS classes

    Returns:
        str: HTML for the button
    """
    # Base classes
    classes = "bg-white border border-gray-300 text-gray-700 hover:bg-gray-50 rounded-lg font-medium transition-colors"

    # Size classes
    size_classes = {
        "sm": "px-3 py-1.5 text-sm",
        "md": "px-4 py-2.5",
        "lg": "px-6 py-3 text-lg"
    }

    classes += f" {size_classes.get(size, size_classes['md'])}"

    # Width class
    if full_width:
        classes += " w-full"

    # Add any additional classes
    if additional_classes:
        classes += f" {additional_classes}"

    # Add icon if provided
    icon_html = f"{icon} " if icon else ""

    return f'<button type="{type}" class="{classes}">{icon_html}{text}</button>'

def link(text, href="#", size="md", icon=None, additional_classes=""):
    """
    Generate HTML for a link styled as a button.

    Args:
        text (str): Button text
        href (str): Link destination
        size (str): Button size (sm, md, lg)
        icon (str): Optional HTML for an icon
        additional_classes (str): Additional CSS classes

    Returns:
        str: HTML for the link button
    """
    # Base classes
    classes = "text-primary hover:text-primary-hover font-medium transition-colors"

    # Size classes
    size_classes = {
        "sm": "text-sm",
        "md": "",
        "lg": "text-lg"
    }

    classes += f" {size_classes.get(size, size_classes['md'])}"

    # Add any additional classes
    if additional_classes:
        classes += f" {additional_classes}"

    # Add icon if provided
    icon_html = f"{icon} " if icon else ""

    return f'<a href="{href}" class="{classes}">{icon_html}{text}</a>'